test/upgrade/autoscaler.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"time"
+
+	"go.uber.org/zap"
+
+	"golang.org/x/sync/errgroup"
+	"knative.dev/serving/test"
+
+	pkgupgrade "knative.dev/pkg/test/upgrade"
+
+	"knative.dev/serving/pkg/apis/autoscaling"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test/e2e"
+)
+
+const (
+	containerConcurrency = 6
+	targetUtilization    = 0.7
+)
+
+// AutoscaleSustainingTest checks that when traffic increases a knative app
+// scales up and sustains the scale as long as the traffic sustains, despite whether
+// it is switching modes between normal and panic.
+func AutoscaleSustainingTest(logger *zap.Logger) pkgupgrade.BackgroundOperation {
+	var ctx *e2e.TestContext
+	var grp errgroup.Group
+	stopCh := make(chan time.Time)
+	return pkgupgrade.NewBackgroundVerification("AutoscaleSustainingTest",
+		func(c pkgupgrade.Context) {
+			// Setup
+			ctx = e2e.SetupSvc(c.T, autoscaling.KPA, autoscaling.Concurrency, containerConcurrency, targetUtilization,
+				rtesting.WithConfigAnnotations(map[string]string{
+					autoscaling.TargetBurstCapacityKey: "0", // Not let Activator in the path.
+				}))
+			grp.Go(func() error {
+				return e2e.AssertAutoscaleUpToNumPods(ctx, logger.Sugar().Infof, 1, 10, stopCh, false /* quick */)
+			})
+		},
+		func(c pkgupgrade.Context) {
+			test.EnsureTearDown(c.T, ctx.Clients(), ctx.Names())
+			// Verification is done inside e2e.AssertAutoscaleUpToNumPods.
+			// We're just giving it a signal.
+			close(stopCh)
+			if err := grp.Wait(); err != nil {
+				c.T.Error("Error: ", err)
+			}
```

Comment:
The whole point of this framework was to simplify those background verification tasks and their synchronization.

Maybe instead of doing here this "walkaround" a better idea would be to split `e2e.AssertAutoscaleUpToNumPods` to two parts - one doing setup, and one doing verification. Removing a stop channel, as there is already one in framework. That should simplify code a lot.

Also, for current code the function `pkgupgrade.NewBackgroundoperation` seams to be better suited, then simplified `pkgupgrade.NewBackgroundVerification`

Reply:
The function `AssertAutoscaleUpToNumPods` is used in E2E tests in 8 other places. Splitting the function would mean more complicated usage in all the places and then maybe we'd save 2-3 lines in this upgrade test. I don't think it's worth it. Better to keep it straightforward in all the other test suites. Also, why do you think NewBackgroundOperation is more suited for this? I don't have any way to define what happens `OnWait` because that's buried in AssertAutoscaleUpToNumPods. I can't see the benefit of changing this. But maybe I'm missing something...

Reply:
Okay. I see.

Other proposal from me: Keep `e2e.AssertAutoscaleUpToNumPods` as it is, but rewrite it have two distinct parts separated with channel. Then use those parts (setup & verify) here, and on other places use `e2e.AssertAutoscaleUpToNumPods`.

Forget about using `NewBackgroundOperation` - you are also invoking, so that appropriate:
```golang
test.EnsureTearDown(c.T, ctx.Clients(), ctx.Names())
```

Reply:
OK. I've pushed the split. It simplifies calling the e2e.AssertAutoscaleUpToNumPods in the other places. And the upgrade tests call the inner parts.

Reply:
Okay. Let's follow up that idea in separate follow up? Could you create an issue for that @mgencur?

Reply:
Yeah. I'll create an issue for this.

Reply:
https://github.com/knative/serving/issues/10276